backend/campaigns/models.py

```
@@ -0,0 +1,288 @@
+def _matches_prefix(path, prefix):
+    return path == prefix or path.startswith(prefix.rstrip('/') + '/')
+
+
+def validate_destination_path(path):
+    """Validate a campaign destination as a safe relative portal path.
+
+    Runs on write (model clean) AND again in the resolver before every
+    redirect, so corrupt stored data fails closed instead of redirecting.
+    """
+    if not isinstance(path, str) or not path:
+        raise ValidationError('Destination is required.')
```

Ruff (0.16.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)